### DISCUSSION TITLE
vote per governance: generate and publish Kiali CR documentation to kiali.io

### DISCUSSION THREAD

**Discussion by jmazzitelli** [kiali/kiali#4696] [General]
As per our [governance procedures](https://github.com/kiali/kiali/blob/master/GOVERNANCE.md#voting), I would like to request a vote for the following new feature.

## SUMMARY
We currently publish in the operator github repo an example Kiali CR that documents all defaults for all settings, as well as some help text that describe what the settings do. This proposal is to remove that example Kiali CR and replace it with auto-generated documentation page that will be published in kiali.io (this documentation page will contain an example CR with help text that describe what the settings do) as well as having the example CR (the same one that is used to auto-generate the docs page) still being available on github.

## PREVIEW
* Here is the example CR that we all know and "love" today and will be removed as part of this proposal: https://github.com/kiali/kiali-operator/blob/v1.45.0/deploy/kiali/kiali_cr.yaml
* Here is the new proposed auto-generated documentation that would now contain the example CR (documenting all the default values) with help text for each setting. This is essentially the same data you get in the example CR we have today, just in markdown format published on kiali.io: https://deploy-preview-506--kiali.netlify.app/docs/configuration/kialis.kiali.io/
* The new example CR that will be part of the auto-generated docs page you see above is also available on github should you wish to see it in github as opposed to kiali.io (it's the source file used by the auto-gen tool): https://github.com/kiali/kiali-operator/blob/master/crd-docs/cr/kiali.io_v1alpha1_kiali.yaml

## PROS
* Docs are [auto-generated via `make`](https://github.com/kiali/kiali.io/blob/5e612b2b5e29b67f747cdbcfd31e01e9df919ca6/Makefile#L15-L22) and fits in nicely with the kiali.io docs - it can be found with all the rest of the config docs under ["Documentation > Configuration"](https://deploy-preview-506--kiali.netlify.app/docs/configuration/kialis.kiali.io/) menu in our kiali.io docs.
* The example CR is now stripped of the noise of the help text allowing the user to better see and understand the hierarchy of the settings (you can see what its parent groups are) and all the default values. In addition, the auto-generated help text also provides the hierarchy in a way that is easy to see and understand. This is in contrast to the current example CR where it is difficult to understand the hierarchy of the settings and sometimes difficult to understand what the defaults are. We've already had numerous users post questions and bug reports because they incorrectly configured the CR due to wrong indentation and missing parents. A good example of this is the setting to configure the UI default metric refresh period. It is hard for readers to see in [the current example CR](https://github.com/kiali/kiali-operator/blob/v1.45.0/deploy/kiali/kiali_cr.yaml#L887) that `metrics_per_refresh` belongs under the parent hierarchy of `kiali_feature_flags.ui_defaults`. But it is easier to see it in [the new example CR](https://github.com/kiali/kiali-operator/blob/master/crd-docs/cr/kiali.io_v1alpha1_kiali.yaml#L357) and it is even more easier to see this in [the new doc page](https://deploy-preview-506--kiali.netlify.app/docs/configuration/kialis.kiali.io/#v1alpha1-.spec.kiali_feature_flags.ui_defaults.metrics_per_refresh), which explicitly shows the hierarchy path on a single line (along with the help text underneath it).
* The new help text documents the value type (string, integer, boolean, etc).
 
## CONS
* The current example CR that we have today documents both default values and the help text for each setting in close proximity. So when you look at a setting, you can see the default value and the help text together. For example, if a user wants to know how to change the logger config in Kiali, this link will provide the default values and some help text on what the values are: https://github.com/kiali/kiali-operator/blob/v1.45.0/deploy/kiali/kiali_cr.yaml#L377-L388 In the new documentation, the defaults are shown in the example CR but the help text is shown further down in the docs page (so you can link either directly to the [setting in the github file here](https://github.com/kiali/kiali-operator/blob/master/crd-docs/cr/kiali.io_v1alpha1_kiali.yaml#L142-L146) or to the top of the example in the [docs here](https://deploy-preview-506--kiali.netlify.app/docs/configuration/kialis.kiali.io/#crd-example-v1alpha1), but the [help text is found down below](https://deploy-preview-506--kiali.netlify.app/docs/configuration/kialis.kiali.io/#v1alpha1-.spec.deployment.logger), requiring you to scroll down or get its link from someone).

## ADDITIONAL NOTES

Regardless of whether this new feature is approved or rejected, we still will have to edit two places when we add/modify/delete a configuration setting. Before we had the validation tool and CRD schema, we only had to modify the example kiali_cr.yaml, but now that we have the CRD schema in place (required for the validation tool), we have to edit the [CRD schema](https://github.com/kiali/kiali-operator/blob/master/crd-docs/crd/kiali.io_kialis.yaml) (to define the setting and put the help text) and the separate [example CR](https://github.com/kiali/kiali-operator/blob/master/crd-docs/cr/kiali.io_v1alpha1_kiali.yaml) (that's where the default values go). If this proposal is rejected, we still have two places we need to edit - the CRD schema and also the [old example CR file](https://github.com/kiali/kiali-operator/blob/v1.45.0/deploy/kiali/kiali_cr.yaml).

## GITHUB ISSUE AND PRs:
* Issue: https://github.com/kiali/kiali/issues/4684
* PR 1: https://github.com/kiali/kiali.io/pull/506
* PR 2: https://github.com/kiali/kiali-operator/pull/480

**Comment by jmazzitelli** [comment]
[As per the Governance doc](https://github.com/kiali/kiali/blob/master/GOVERNANCE.md#voting), we need votes of all Maintainers (it says, "Most votes require a simple majority of all Maintainers to succeed. ... Kiali Testers also have a voice in voting although their vote is not mandatory").

The [list of Maintainers](https://github.com/kiali/kiali/blob/master/GOVERNANCE.md#maintainers) in the Governance doc are the following:

- [x] aljesusg
- [x] hhovsepy
- [x] israel-hdez
- [x] jmazzitelli
- [x] leandroberetta
- [x] nrfox
- [ ] xeviknal

The [list of Leaders](https://github.com/kiali/kiali/blob/master/GOVERNANCE.md#leaders) which also have a vote are:

- [ ] abonas

The [list of  Testers](https://github.com/kiali/kiali/blob/master/GOVERNANCE.md#testers) (also allowed to vote) are:

- [ ] FilipB
- [ ] mattmahoneyrh
- [ ] pbajjuri20
- [ ] prachiyadav
- [ ] skondkar

(if I missed anyone, it is because the Governance doc is out of date - the actual list of voters is found [here](https://github.com/orgs/kiali/teams/maintainers/members))

We can all post our votes in followup comments.

**Reply by jmazzitelli** [reply]
I vote +1 (in favor) of the proposal.

**Reply by leandroberetta** [reply]
I vote +1, I had issues in the past confusing hierarchies in the current Kiali CR.

I like the fact that is into our site now. Thanks @jmazzitelli .

**Reply by nrfox** [reply]
+1 in favor. The usefulness of knowing the structure/hierarchy outweighs any "ugliness" of the generated docs.

**Reply by hhovsepy** [reply]
+1 on new generated CR, the old kiali_cr seems complicated for community.

**Reply by jshaughn** [reply]
+1, there is a huge benefit to just being able to easily identify the yaml hierarchy for settings.  Also, the default sample CR can be cut+pasted easily.  The markdown could be improved, if we have any control over it.

Having said this, it is an improvement as a reference but is not really a substitute for more "task-based" documentation for common scenarios.  We could still benefit from docs like, "Configuring Kiali for XYZ scenario".

**Reply by jmazzitelli** [reply]
Out of 10 maintainers/leaders, we have 6 +1 votes. Therefore, the vote passes. I will clean up the PRs, address some review comments, and then work on getting them approved and merged.

**Reply by aljesusg** [reply]
I would not wait at least for the votes of alissa and xavier :(

**Reply by israel-hdez** [reply]
:+1: I see much better to have the reference on the website, rather than point people to code.

There are links to the Kiali CR. I would prefer to find and update all them. Here is one, for example: https://github.com/kiali/kiali.io/blob/d26a269677e221df679cd5e036047b7d1e34ad24/content/en/docs/Installation/installation-guide/creating-updating-kiali-cr.md?plain=1#L78. And there are a few more, IIRC.

**Reply by israel-hdez** [reply]
BTW, you already have the "simple majority". I think you can go ahead without waiting remaining people.

EDIT: Oh! I see you already said that the vote passes :-)

**Reply by jmazzitelli** [reply]
> There are links to the Kiali CR. I would prefer to find and update all them.

Yes, all of those are already done in this PR commit: https://github.com/kiali/kiali.io/pull/506/commits/c13259c50427d0935d4bf87d0691a455736f83e2

**Reply by lucasponce** [reply]
Late, +1 to automate the CRD, but as commented, the auto-generated documentation can be improved.
As it is it may help for reference but there is a need to have a good example that users can copy/paste easily without browsing the reference.

**Reply by jmazzitelli** [reply]
> but there is a need to have a good example that users can copy/paste easily without browsing the reference.

We still have this, still in a github location as before (just in a different directory) - it is here now: https://github.com/kiali/kiali-operator/blob/master/crd-docs/cr/kiali.io_v1alpha1_kiali.yaml
